packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.ts

```diff
@@ -0,0 +1,279 @@
+  const mappedUserActions = rawActions.map((rawAction) => {
+    const actionData = JSON.parse(rawAction.value);
+    return {
+      actionType: rawAction.name as UserActionType,
+      timestamp: rawAction.time,
+      eventData: {
+        type: actionData.event_type,
+        value: actionData.event_value,
+      },
+      cause: actionData.cause,
+      searchHub: actionData.origin_level_1,
+      document: {
+        title: actionData.title,
+        uriHash: actionData.uri_hash,
+        // eslint-disable-next-line @cspell/spellchecker
+        contentIdKey: actionData.c_contentidkey,
+        // eslint-disable-next-line @cspell/spellchecker
+        contentIdValue: actionData.c_contentidvalue,
+      },
+      query: actionData.query_expression,
+    };
+  });
+  return mappedUserActions;
+};
+
+export const splitActionsIntoTimelineSessions = (
+  actions: UserAction[],
+  ticketCreationDate: string
+): UserActionTimeline => {
+  const returnTimeline: UserActionTimeline = {
+    precedingSessions: [],
+    session: undefined,
+    followingSessions: [],
+    caseSessionFound: false,
+  };
+
+  let currentSession: UserSession = {
+    start: actions[0].timestamp,
+    end: actions[0].timestamp,
+    actions: [],
+  };
+
+  actions.forEach((action) => {
+    if (isPartOfTheSameSession(action, currentSession.end)) {
+      currentSession.actions.push(action);
+      currentSession.start = action.timestamp;
+      return;
+    }
+
+    handleSessionEnd(currentSession, returnTimeline, ticketCreationDate);
+
+    currentSession = {
+      start: action.timestamp,
+      end: action.timestamp,
+      actions: [action],
+    };
+  });
+
+  // Inserting ticket creation action in current session
+  const caseCreationSessionActions: UserAction[] | [] =
+    insertTicketCreationActionInSession(
+      returnTimeline.session,
+      ticketCreationDate
+    );
+
+  const caseCreationSession = createCaseCreationSession(
+    caseCreationSessionActions
+  );
+  returnTimeline.session = caseCreationSession;
+
+  return returnTimeline;
+};
+
+// Checks if an action is within 30mins of the previous action
+export const isPartOfTheSameSession = (
+  action: UserAction,
+  previousEndDateTime: string
+): boolean => {
+  return (
+    Math.abs(Number(action.timestamp) - Number(previousEndDateTime)) <
+    MAX_MSEC_IN_SESSION
+  );
+};
+
+// Handles the end of a session
+const handleSessionEnd = (
+  currentSession: UserSession,
+  returnTimeline: UserActionTimeline,
+  ticketCreationDate: string
+) => {
+  if (currentSessionHasTicketCreation(currentSession, ticketCreationDate)) {
+    returnTimeline.session = currentSession;
+    returnTimeline.caseSessionFound = true;
+    return;
+  }
+
+  if (currentSessionIsBeforeCaseCreation(currentSession, ticketCreationDate)) {
+    returnTimeline.precedingSessions.push(currentSession);
+    return;
+  }
+
+  returnTimeline.followingSessions.push(currentSession);
+};
```

Comment:
One thing to keep in mind, dividing  into smaller functions is not always something that will reduce complexity.

For me this function is not reducing complexity cause it's not a standalone function with one sole objective, looking at it alone will not make much sense for me.

Reply:
It's both because of what it does, and because of its name.

for example I would keep these smaller divisions, but change the names a bit:

```
currentSessionHasTicketCreation -> isTimestampInSession(timestamp, session);
currentSessionIsBeforeCaseCreation -> isTimestampAfterSession(timestamp, session);
currentSessionIsAfterCaseCreation -> isTimestampBeforeSession(timestamp, session);
```

```suggestion
const placeSessionInTimeline = (
  session: UserSession,
  timeline: UserActionTimeline,
  ticketCreationDate: string
) => {
  if (isTimestampInSession(ticketCreationDate, session)) {
    timeline.caseSession = currentSession;
  } else if (isTimestampAfterSession(ticketCreationDate, session)) {
    timeline.precedingSessions.push(session);
  } else {
    timeline.followingSession.push(session);
  }
};
```